survey_app/src/main/java/org/opendatakit/survey/fragments/FormChooserListFragment.java

```diff
@@ -102,17 +177,79 @@
 
   @Override public void onLoadFinished(Loader<ArrayList<FormInfo>> loader,
       ArrayList<FormInfo> dataset) {
+
+    String sortingOrder = mPropSingleton.getProperty(KEY_SURVEY_SORT_ORDER);
+    mItems = dataset;
+    sortFormList(dataset,sortingOrder);
+
     // Swap the new cursor in. (The framework will take care of closing the
     // old cursor once we return.)
-    mAdapter.clear();
-    mAdapter.addAll(dataset);
-    mAdapter.notifyDataSetChanged();
+    mAdapter.swapData(dataset);
   }
 
+
   @Override public void onLoaderReset(Loader<ArrayList<FormInfo>> loader) {
     // This is called when the last Cursor provided to onLoadFinished()
     // above is about to be closed. We need to make sure we are no
     // longer using it.
     mAdapter.clear();
   }
+
+  //  Sorts the forms list according to sorting order
+  private void sortFormList(ArrayList<FormInfo> forms, String sortingOrder) {
+
+    if (sortingOrder.equals(SORT_BY_NAME)) {
+      Collections.sort(forms, new Comparator<FormInfo>() {
+        @Override
+        public int compare(FormInfo lhs, FormInfo rhs) {
+          int cmp = lhs.formDisplayName.compareToIgnoreCase(rhs.formDisplayName);
+          if (cmp != 0) {
+            return cmp;
+          }
+          cmp = lhs.tableId.compareToIgnoreCase(rhs.tableId);
+          if (cmp != 0) {
+            return cmp;
+          }
+          cmp = lhs.formId.compareToIgnoreCase(rhs.formId);
+          if (cmp != 0) {
+            return cmp;
+          }
+          cmp = lhs.formVersion.compareToIgnoreCase(rhs.formVersion);
+          if (cmp != 0) {
+            return cmp;
+          }
+          cmp = lhs.formDisplaySubtext.compareToIgnoreCase(rhs.formDisplaySubtext);
+          return cmp;
+        }
+      });
+    } else if (sortingOrder.equals(SORT_BY_TABLEID)) {
+
+      Collections.sort(mItems, new Comparator<FormInfo>() {
```

Comment:
It should sort the method parameter `forms` not the instance variable `mItems`. They behave identically here but future changes could break it.

Reply:
Good catch

Reply:
@linl33 thanks for this :+1: